File: zero_matrix/zero_matrix.cpp
/*
    Write an algorithm such that if an element in an `M x N` matrix is `0`, its entire row and column are set to `0`.
*/

#include <iostream>
#include <vector>
#include <tuple>

using namespace std;

void zero_matrix(vector< vector<int> > &matrix) {
    vector< tuple<int, int> > zero_pos;

    for (int m = 0; m < matrix.size(); m++) {
        for (int n = 0; n < matrix[m].size(); n++) {
            if (matrix[m][n] == 0) {
                zero_pos.push_back(make_tuple(m, n));
            }
        }
    }

    for (int i = 0; i < zero_pos.size(); i++) {
        for (int j = 0; j < matrix.size(); j++) {
            matrix[get<0>(zero_pos[i])][j] = 0;
        }

        for (int k = 0; k < matrix[get<0>(zero_pos[i])].size(); k++) {
            matrix[k][get<1>(zero_pos[i])] = 0;
        }
    }
}

int main() {
    vector< vector<int> > matrix;
    vector<int> aux;

    aux.push_back(1);
    aux.push_back(2);
    aux.push_back(3);
    aux.push_back(4);

    matrix.push_back(aux);
    aux.clear();

    aux.push_back(5);
    aux.push_back(6);
    aux.push_back(0);
    aux.push_back(8);

    matrix.push_back(aux);
    aux.clear();

    aux.push_back(9);
    aux.push_back(10);
    aux.push_back(11);
    aux.push_back(12);

    matrix.push_back(aux);
    aux.clear();

    aux.push_back(13);
    aux.push_back(14);
    aux.push_back(15);
    aux.push_back(16);

    matrix.push_back(aux);
    aux.clear();

    for (int i = 0; i < matrix.size(); i++) {
        for (int j = 0; j < matrix[i].size(); j++) {
            cout << matrix[i][j] << "\t";
        }

        cout << endl;
    }

    zero_matrix(matrix);

    cout << "-----------------------------------" << endl;

    for (int i = 0; i < matrix.size(); i++) {
        for (int j = 0; j < matrix[i].size(); j++) {
            cout << matrix[i][j] << "\t";
        }

        cout << endl;
    }
}

File: zero_matrix/zero_matrix.py
"""
    Write an algorithm such that if an element in an `M x N` matrix is `0`, its entire row and column are set to `0`.
"""


def zero_matrix(matrix):
    """
    Transform row and column into 0 if there is a 0 in them.

    Parameters
    ----------
    matrix : List
        Matrix to transform.
    """
    zero_pos = []

    for m in range(0, len(matrix)):
        for n in range(0, len(matrix[m])):
            if matrix[m][n] == 0:
                zero_pos.append((m, n))

    for pos in zero_pos:
        for i in range(0, len(matrix)):
            matrix[pos[0]][i] = 0

        for j in range(0, len(matrix[pos[0]])):
            matrix[j][pos[1]] = 0

    return matrix


mat = [[1, 2, 3, 4], [5, 6, 0, 8], [9, 10, 11, 12], [13, 14, 15, 16]]

for elem in mat:
    print(elem)

new_mat = zero_matrix(mat)

print("\n---------------\n")

for elem in new_mat:
    print(elem)
